### DISCUSSION TITLE
Install Verdaccio into a monorepo

### DISCUSSION THREAD

**Discussion by adamSneller** [verdaccio/verdaccio#3232] [Q&A]
I would like to install Verdaccio into my monorepo, so that whenever I execute `git push`, my local npm registry gets backed, up along with my source code to my repo on Github. Notionally, this would be configured as:

    monorepo
      - components
      - verdaccio
         - database
         - storage
         - package.json

How can I configure Verdaccio so that when I call it from within the monorepo, it knows that it should use the `database` and `storage` located within the `verdaccio` workspace?

**Comment by edclement** [comment]
Have you looked at the docs for configuring?
* https://verdaccio.org/docs/cli#commands
* https://verdaccio.org/docs/configuration#storage

**Reply by adamSneller** [reply]
Thanks @edclement! I managed to figure this out. It actually works perfectly. The only thing that I'm still stumped on is how to get Verdaccio to display the correct `publisher` (the user who actually publishes the version to the registry) instead of the `author`, which appears to be hard-coded to each package's `package.json`? for example:

    {
      "name": "@ms2d/test-component",
      "version": "1.0.0",
      "author": {
          "name": "Author Name",
          "email": "author@someco.com"
      },
      "main": "index.js",
      "type": "module"
    }

**Reply by adamSneller** [reply]
So if anyone else is looking for a solution to using Verdaccio within a Lerna/NPM8 monorepo, here is what I have so far:

Folder structure
---------------

    - packages
       - test-component
          - index.js
          - package.json
    - verdaccio
       - plugins
       - storage
       - config.yaml
       - htpasswd
       - package.json
    - .gitignore
    - lerna.json
    - package.json

Config Files
------------

**/package.json**

    {
      "name": "@xyzco/root",
      "private": true,
      "devDependencies": {
        "lerna": "^4.0.0"
      },
      "scripts": {
        "verdaccio": "cd verdaccio && verdaccio --config ./config.yaml"
      }
    }

**/lerna.json**

    {
      "packages": [
        "packages/*"
      ],
      "version": "independent",
      "command": {
        "publish": {
          "registry": "http://localhost:4873"
        }
      }
    }

**/.gitignore**

    node_modules

**/verdaccio/config.yaml**

    storage: ./storage
    plugins: ./plugins
    auth:
      htpasswd:
        file: ./htpasswd
    web:
      enable: true
      title: XYZCO
      logo: ./logo.png
      primary_color: "#ffffff"
    uplinks:
      npmjs:
        url: https://registry.npmjs.org/
    packages:
      "@*/*":
        access: $all
        publish: $authenticated
        proxy: npmjs
      "**":
        proxy: npmjs
    log: { type: stdout, format: pretty, level: http }

**/packages/test-component/package.json**

    {
      "name": "@xyzco/test-component",
      "version": "1.0.0",
      "author": {
        "name": "Admin",
        "email": "admin@xyzco.com"
     },
      "main": "index.js",
      "type": "module"
    }

Installation
-----------

So commands to set this up would be something like:

    mkdir test && cd test
    npx lerna init --independent

Create folders and config files (above). Then from the root of your monorepo run:

    npm install
    npm install -g verdaccio
    npm adduser --registry==http://localhost:4873
    npm run verdaccio
    git add .
    git commit -m "initial commit"
    git push
    lerna publish

If all goes well, this should show up in Verdaccio!
